### DISCUSSION TITLE
Installing private dependencies with yarn via SSH fails, while npm works

### DISCUSSION THREAD

**Discussion by vecuenca** [community/community#26472] [Actions]
<p>I’m working with two private organizations. One of the repos in one org has a dependency on a package in the other org.</p>
<p>I’m trying to migrate this repo from npm to yarn, and have updated the workflow like so:</p>
<pre><code>jobs:
  build:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [10.x]

    steps:
      - uses: actions/checkout@v1

      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - uses: webfactory/ssh-agent@v0.2.0
        with:
          ssh-private-key: ${{ secrets.private_ssh }}

      - name: Install dependencies
        run: yarn install
</code></pre>
<p>This fails with:</p>
<pre><code>Run yarn install
  yarn install
  shell: /bin/bash -e {0}
  env:
    SSH_AUTH_SOCK: /tmp/ssh-auth.sock
    CI: true
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
error Command failed.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Exit code: 128
Command: git
Arguments: ls-remote --tags --heads ssh://git@github.com/org/repo.git
Directory: /home/runner/work/repo/repo
Output:
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
##[error]Process completed with exit code 128.
</code></pre>
<p>The previous version of this workflow (using npm i instead of the yarn install) works, and both yarn and npm install dependencies on my local machine.</p>
<p>Any ideas on how to resolve this would be welcome :slight_smile:</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/vecuenca">@vecuenca</a> , </p>
<p>What’s the syntax for private github repo in dependencies section of package.json file?</p>
<p>Please try the next format: </p>
<pre><code>"dependencies": {
    "&lt;PackageName&gt;": "git+ssh://git@github.com:\&lt;user\&gt;/\&lt;repo\&gt;.git",
    "foo" : "1.0.0 - 2.9999.9999"
  }
</code></pre>
<p> This works fine in my side. Please kindly let me know your result. </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/3/38d312c8395d050f19454385cb9a3b9e7e793b73.png" data-download-href="/uploads/short-url/86GXzTeD0VIarO5egMKvx1BzKN5.png?dl=1" title="yarn install.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181097938-f1562dce-0c12-483e-afd0-2942e05297d2.png" alt="yarn install.png" data-base62-sha1="86GXzTeD0VIarO5egMKvx1BzKN5" width="524" height="499" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/38d312c8395d050f19454385cb9a3b9e7e793b73_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">yarn install.png</span><span class="informations">650×620 13.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by vecuenca** [comment]
<p>Hello <a class="mention" href="/yanjingzhu">@yanjingzhu</a>, thank you for your response</p>
<p>I have two references to the private github repo, both are in the format</p>
<pre><code>"&lt;package-name&gt;": "git+ssh://git@github.com/&lt;org&gt;/&lt;repo&gt;.git",
</code></pre>
<p>I changed the slash after the domain name to a colon as per your example and it’s still failing with:</p>
<pre><code>Output:
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: Command failed.
Exit code: 128
Command: git
Arguments: ls-remote --tags --heads git@github.com:org/repo.git
Directory: /home/runner/work/Repo/Repo
Output:
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
    at ProcessTermError.ExtendableBuiltin (/usr/share/yarn/lib/cli.js:721:66)
    at ProcessTermError.MessageError (/usr/share/yarn/lib/cli.js:750:123)
    at new ProcessTermError (/usr/share/yarn/lib/cli.js:790:113)
    at ChildProcess.&lt;anonymous&gt; (/usr/share/yarn/lib/cli.js:25884:17)
    at ChildProcess.emit (events.js:198:13)
    at maybeClose (internal/child_process.js:982:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
Error: Command failed.
Exit code: 128
</code></pre>

**Comment by vecuenca** [comment]
<p>Another attempt at solving this:</p>
<p>I’ve generated two new ssh keys and placed the .pub files as deploy keys for the two external repos. I’ve then set the private keys as secrets in the repository. The PR workflow YAML has been updated to this:</p>
<pre><code>steps:
      - uses: actions/checkout@v1

      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - uses: webfactory/ssh-agent@v0.2.0
        with:
          ssh-private-key: |
            ${{ secrets.repo1_deploy_key }}
            ${{ secrets.repo2_deploy_key }}

      - name: Install dependencies
        run: yarn install
        env:
          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
</code></pre>
<p>I’ve set SSH_AUTH_SOCK to the value defined here (<a href="https://www.webfactory.de/blog/use-ssh-key-for-private-repositories-in-github-actions" rel="nofollow noopener">https://www.webfactory.de/blog/use-ssh-key-for-private-repositories-in-github-actions</a>).</p>
<p>This now results in</p>
<pre><code>Run yarn install
  yarn install
  shell: /bin/bash -e {0}
  env:
    SSH_AUTH_SOCK: /tmp/ssh_agent.sock
yarn install v1.22.4
[1/4] Resolving packages...
[2/4] Fetching packages...
error Command failed.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Exit code: 128
Command: git
Arguments: ls-remote --tags --heads git@github.com:&lt;org&gt;/&lt;repo&gt;.git
Directory: /home/runner/work/&lt;org&gt;/&lt;org&gt;
Output:
Warning: Permanently added the RSA host key for IP address '140.82.114.4' to the list of known hosts.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
##[error]Process completed with exit code 128.
</code></pre>
<p>(whereas not specifying SSH_AUTH_SOCK results in the original repository not found error.)</p>

**Comment by Yanjingzhu** [comment]
<p>Are your SSH private key right ones? To troubleshoot your issue, would you mind changing to <strong>npm install</strong> and ensure npm install could work with ssh ? </p>

**Comment by vecuenca** [comment, marked as answer]
<p>I found the issue. The SSH key we were using was out of date.</p>
<p>Although I’m not sure why the new deploy keys I created didn’t work.</p>
<p>Thank you for your assistance.</p>

**Comment by CezaryDanielNowak** [comment]
I had similar issue and posted a fix here https://github.com/yarnpkg/yarn/issues/2614#issuecomment-1543917535
